### ISSUE TITLE
Support tagged-union narrowing via `isinstance` checks

### ISSUE THREAD

**Issue by zeevro** [astral-sh/ty#4376]
### Summary

`ty` behaves weirdly when trying to narrow `TypedDict`s with `closed=True`:

```py
from typing import reveal_type

from typing_extensions import TypedDict


class A(TypedDict, closed=True):
    id: int
    body: list[tuple[int, str]]


class B(TypedDict, closed=True):
    id: int
    body: bytes


type X = A | B


def f() -> X:
    return {'id': 0, 'body': b''}


x = f()

reveal_type(x['body'])  # list[tuple[int, str]] | bytes - correct
reveal_type(x)  # A | B - correct

if isinstance(x['body'], list):
    reveal_type(x)  # A | B - should be A
elif isinstance(x['body'], str):
    reveal_type(x)  # A | B - should be Never
elif isinstance(x['body'], int):
    reveal_type(x)  # A | B - should be Never
elif isinstance(x['body'], bytes):
    reveal_type(x)  # A | B - should be B
elif isinstance(x['body'], str):
    reveal_type(x)  # Never - correct
elif isinstance(x['body'], int):
    reveal_type(x)  # Never - correct
else:
    reveal_type(x)  # Never - correct


if isinstance(x['body'], list):
    reveal_type(x['body'])  # list[tuple[int, str]] - correct
    reveal_type(x)  # A | B - should be A
elif isinstance(x['body'], str):
    reveal_type(x['body'])  # Never - correct
    reveal_type(x)  # Never - correct
elif isinstance(x['body'], int):
    reveal_type(x['body'])  # Never - correct
    reveal_type(x)  # Never - correct
elif isinstance(x['body'], bytes):
    reveal_type(x['body'])  # bytes  - correct
    reveal_type(x)  # A | B - should be B
elif isinstance(x['body'], str):
    reveal_type(x['body'])  # Never - correct
    reveal_type(x)  # Never - correct
elif isinstance(x['body'], int):
    reveal_type(x['body'])  # Never - correct
    reveal_type(x)  # Never - correct
else:
    reveal_type(x['body'])  # Never - correct
    reveal_type(x)  # Never - correct
```
https://play.ty.dev/54d03691-342c-4ee1-aba1-001dc14c116e

Not only is there wrong (and weird) type-narrowing here, there's also some weird code reachability conclusions (`ty` decides code is unreachable mid-branch, only after `x['body']` is mentioned)

### Version

ty 0.0.74 (00199f0aa 2026-08-22)

**Comment by sharkdp** [comment]
> Not only is there wrong (and weird) type-narrowing here, there's also some weird code reachability conclusions (`ty` decides code is unreachable mid-branch, only after `x['body']` is mentioned)

Those reachability conclusions are not incorrect. `reveal_type` has type `reveal_type[T](typ: T) -> T`, just like a generic identity function. So if you pass in something of type `Never`, you get `Never` as a return type -- and nothing after that `reveal_type` call can be reached.

It's better to put the examples in separate scopes to prevent this.

**Comment by carljm** [comment]
Hi, thanks for the report!

It seems like the core of this report is not really a bug, just a feature request for tagged-union narrowing -- we don't support tagged union narrowing via `isinstance` checks, only via comparisons to literal types. Some other type checkers support tagged-union narrowing via `isinstance`, not all (pyright does not). It's a bit subtle because it presents another case where we have to decide whether to respect the possibility of multiple inheritance or not. But this is a reasonable feature request, and we can use this issue to track it.

I'm not sure if there are any other "buggy behaviors" observed in this example that aren't explained either by that missing feature, or by the `reveal_type` behavior described by @sharkdp . (Maybe we should special-case `reveal_type` so it is never treated as terminal?) If so, it would be most useful to isolate them in their own issue.
